File: MachineLearning/GeneticAlgorithms/mutations/mutations.go
package main

import (
	"math/rand"
	"time"
)

// BitFlipMutation
func BitFlipMutation(chromosome []int, mutationRate float64) []int {
	for i := 0; i < len(chromosome); i++ {
		if rand.Float64() < mutationRate { // Chance of mutation
			chromosome[i] = 1 - chromosome[i] // Flip the bit
		}
	}
	return chromosome
}

// SwapMutation
func SwapMutation(chromosome []int, mutationRate float64) []int {
	if rand.Float64() < mutationRate {
		i, j := rand.Intn(len(chromosome)), rand.Intn(len(chromosome))
		for i == j { // Ensure i != j
			j = rand.Intn(len(chromosome))
		}
		chromosome[i], chromosome[j] = chromosome[j], chromosome[i] // Swap values
	}
	return chromosome
}

// ScrambleMutation
func ScrambleMutation(chromosome []int, mutationRate float64) []int {
	if rand.Float64() < mutationRate {
		start, end := rand.Intn(len(chromosome)), rand.Intn(len(chromosome))
		if start > end {
			start, end = end, start
		}
		scrambledSublist := append([]int{}, chromosome[start:end]...)
		rand.Shuffle(len(scrambledSublist), func(i, j int) {
			scrambledSublist[i], scrambledSublist[j] = scrambledSublist[j], scrambledSublist[i]
		})
		copy(chromosome[start:end], scrambledSublist)
	}
	return chromosome
}

// InversionMutation
func InversionMutation(chromosome []int, mutationRate float64) []int {
	if rand.Float64() < mutationRate {
		start, end := rand.Intn(len(chromosome)), rand.Intn(len(chromosome))
		if start > end {
			start, end = end, start
		}
		for i, j := start, end-1; i < j; i, j = i+1, j-1 {
			chromosome[i], chromosome[j] = chromosome[j], chromosome[i] // Reverse the selected range
		}
	}
	return chromosome
}

// GaussianMutation
func GaussianMutation(chromosome []float64, mutationRate float64, sigma float64) []float64 {
	for i := 0; i < len(chromosome); i++ {
		if rand.Float64() < mutationRate {
			chromosome[i] += rand.NormFloat64() * sigma // Add Gaussian noise
		}
	}
	return chromosome
}

// UniformMutation
func UniformMutation(chromosome []float64, mutationRate float64, valueRange [2]float64) []float64 {
	for i := 0; i < len(chromosome); i++ {
		if rand.Float64() < mutationRate {
			chromosome[i] = valueRange[0] + rand.Float64()*(valueRange[1]-valueRange[0]) // Random value in range
		}
	}
	return chromosome
}

// NonUniformMutation
func NonUniformMutation(chromosome []float64, mutationRate float64, generation, maxGenerations int, valueRange [2]float64) []float64 {
	for i := 0; i < len(chromosome); i++ {
		if rand.Float64() < mutationRate {
			delta := (valueRange[1] - valueRange[0]) * (1 - float64(generation)/float64(maxGenerations))
			chromosome[i] += rand.Float64()*2*delta - delta // Apply shrinking delta
			// Ensure the gene stays within the value range
			if chromosome[i] < valueRange[0] {
				chromosome[i] = valueRange[0]
			} else if chromosome[i] > valueRange[1] {
				chromosome[i] = valueRange[1]
			}
		}
	}
	return chromosome
}

// BoundaryMutation
func BoundaryMutation(chromosome []float64, mutationRate float64, valueRange [2]float64) []float64 {
	for i := 0; i < len(chromosome); i++ {
		if rand.Float64() < mutationRate {
			if rand.Float64() < 0.5 {
				chromosome[i] = valueRange[0]
			} else {
				chromosome[i] = valueRange[1]
			} // Choose boundary value
		}
	}
	return chromosome
}

// ArithmeticMutation
func ArithmeticMutation(chromosome []float64, mutationRate float64) []float64 {
	for i := 0; i < len(chromosome); i++ {
		if rand.Float64() < mutationRate {
			chromosome[i] += chromosome[i] * (rand.Float64()*0.2 - 0.1) // Adjust by small percentage
		}
	}
	return chromosome
}

// RandomResettingMutation
func RandomResettingMutation(chromosome []float64, mutationRate float64, valueRange [2]float64) []float64 {
	for i := 0; i < len(chromosome); i++ {
		if rand.Float64() < mutationRate {
			chromosome[i] = valueRange[0] + rand.Float64()*(valueRange[1]-valueRange[0]) // Reset to random value
		}
	}
	return chromosome
}

func main() {
	// Example to initialize the random seed and call mutations
	rand.Seed(time.Now().UnixNano())
	chromosome := []int{1, 0, 1, 0, 1, 0}
	mutationRate := 0.1

	// Test mutation functions
	chromosome = BitFlipMutation(chromosome, mutationRate)
	// Add other mutation functions as needed to test

	// Print the mutated chromosome
	for _, gene := range chromosome {
		print(gene, " ")
	}
}

File: MachineLearning/GeneticAlgorithms/mutations/mutations.py
import random


#Bit Flip
def bit_flip_mutation(chromosome, mutation_rate):
    for i in range(len(chromosome)):
        if random.random() < mutation_rate:  # Chance of mutation
            chromosome[i] = 1 - chromosome[i]  # Flip the bit
    return chromosome

#Swap Mutation
def swap_mutation(chromosome, mutation_rate):
    if random.random() < mutation_rate:
        i, j = random.sample(range(len(chromosome)), 2)  # Select two random indices
        chromosome[i], chromosome[j] = chromosome[j], chromosome[i]  # Swap values
    return chromosome

#Scramble Mutation
def scramble_mutation(chromosome, mutation_rate):
    if random.random() < mutation_rate:
        start, end = sorted(random.sample(range(len(chromosome)), 2))  # Select a range
        scrambled_sublist = chromosome[start:end]
        random.shuffle(scrambled_sublist)  # Shuffle the selected sublist
        chromosome[start:end] = scrambled_sublist  # Place the scrambled sublist back
    return chromosome

#Inversion Mutation
def inversion_mutation(chromosome, mutation_rate):
    if random.random() < mutation_rate:
        start, end = sorted(random.sample(range(len(chromosome)), 2))  # Select a range
        chromosome[start:end] = chromosome[start:end][::-1]  # Reverse the selected range
    return chromosome

#Gaussian Mutation
def gaussian_mutation(chromosome, mutation_rate, sigma=0.1):
    for i in range(len(chromosome)):
        if random.random() < mutation_rate:
            chromosome[i] += random.gauss(0, sigma)  # Add Gaussian noise
    return chromosome

#Uniform Mutation
def uniform_mutation(chromosome, mutation_rate, value_range):
    for i in range(len(chromosome)):
        if random.random() < mutation_rate:
            chromosome[i] = random.uniform(*value_range)  # Assign a random value from the range
    return chromosome

#Non-Uniform Mutation
def non_uniform_mutation(chromosome, mutation_rate, generation, max_generations, value_range):
    for i in range(len(chromosome)):
        if random.random() < mutation_rate:
            delta = (value_range[1] - value_range[0]) * (1 - generation / max_generations)
            chromosome[i] += random.uniform(-delta, delta)  # Apply a shrinking delta over generations
            # Ensure the gene stays within the value range
            chromosome[i] = max(min(chromosome[i], value_range[1]), value_range[0])
    return chromosome

#Boundary Mutation
def boundary_mutation(chromosome, mutation_rate, value_range):
    for i in range(len(chromosome)):
        if random.random() < mutation_rate:
            chromosome[i] = random.choice([value_range[0], value_range[1]])  # Choose a boundary value
    return chromosome

#Arithmetic Mutation
def arithmetic_mutation(chromosome, mutation_rate):
    for i in range(len(chromosome)):
        if random.random() < mutation_rate:
            chromosome[i] += chromosome[i] * random.uniform(-0.1, 0.1)  # Adjust by a small percentage
    return chromosome

#Random Resetting
def random_resetting_mutation(chromosome, mutation_rate, value_range):
    for i in range(len(chromosome)):
        if random.random() < mutation_rate:
            chromosome[i] = random.uniform(*value_range)  # Reset to a random value within the range
    return chromosome